src/unistd.rs

```diff
@@ -2136,3 +2146,424 @@ mod setres {
+/// should be more than adequate. Only use the buffer size variants if you receive an ERANGE error
+/// without them, or you know that you will be likely to.
+///
+/// # Examples
+///
+/// ```
+/// use nix::unistd::{User, UserQuery, Queryable};
+/// let find = String::from("root");
+/// // Returns an Option<Result<User>>, thus the double unwrap.
+/// let res = User::query( UserQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+pub enum UserQuery {
+    /// Get a user by UID. 
+    ///
+    /// Internally, this function calls
+    /// [getpwuid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Uid(Uid),
+    /// Get a user by name. 
+    ///
+    /// Internally, this function calls
+    /// [getpwnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    UidWithBufsize(Uid, usize),
+    NameWithBufsize(String, usize)
+}
+
+/// Query a group. 
+///
+/// The buffer size variants should not be needed 99% of the time; the default buffer size of 1024
+/// should be more than adequate. Only use the buffer size variants if you receive an ERANGE error
+/// without them, or you know that you will be likely to.
+///
+/// # Examples
+///
+/// ```ignore
+/// use nix::unistd::{Group, GroupQuery};
+/// // On many systems there's no `root` group; on FreeBSD for example it's known as `wheel`
+/// let find = String::from("root");
+/// // Returns an Option<Result<Group>>, thus the double unwrap. Will panic if there's no `root`
+/// // group.
+/// let res = Group::query( GroupQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+#[derive(Debug)]
+pub enum GroupQuery {
+    /// Get a group by GID. 
+    ///
+    /// Internally, this function calls
+    /// [getgrgid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Gid(Gid),
+    /// Get a group by name. 
+    ///
+    /// Internally, this function calls
+    /// [getgrnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    GidWithBufsize(Gid, usize),
+    NameWithBufsize(String, usize)
+}
+
+pub trait Queryable<Q> where Self: ::std::marker::Sized {
+    fn query(q: Q) -> Option<Result<Self>>;
+}
+
+/// Default buffer size for system user and group querying functions
+const PWGRP_BUFSIZE: usize = 1024;
+
+impl Queryable<UserQuery> for User {
+    fn query(q: UserQuery) -> Option<Result<Self>> {
+        let mut cbuf = match q {
+            UserQuery::UidWithBufsize(_, size) | 
+            UserQuery::NameWithBufsize(_, size) => {
+                vec![0 as c_char; size]
+            },
+            _ => {
+                vec![0 as c_char; PWGRP_BUFSIZE]
+            }
+        };
+
+        let mut pwd: libc::passwd =  unsafe { mem::zeroed() };
+        let mut res = ptr::null_mut();
+
+        let i = { 
```

Comment:
Can you use a more descriptive variable name here? `i` is almost always a counter variable as well, so this is confusing.

Reply:
changed `i` to `error` for all four occurences